File: Problem_Solving/leetcode/string_control/Most_Common_Word_819/Most_Common_Word.py
import collections
import re

def mostCommonWord(self, paragraph: str, banned: list[str]) -> str:
    # Data Cleansing : 입력값에 대한 전처리 작업
    # 정규식에서 \w는 단어 문자(Word Character)를 뜻하며, ^은 not을 의미한다.
    # 즉 단어 문자가 아닌 모든 문자를 공백으로 치환
    # List Comprehension에서 words에 소문자, 구두점, banned을 제외한 단어 목록을 저장
    words = [word for word in re.sub(r'[\w]', ' ', paragraph)
            .lower().split()
                if word not in banned]
    
    counts = collections.Counter(words)
    
    # 가장 흔하게 등장하는 단어의 첫 번쨰 인덱스 리턴
    # ex) [('ball', 2)]가 되면 [0][0]으로 첫번째 인덱스의 키를 추출
    return counts.most_common(1)[0][0]

File: Problem_Solving/leetcode/string_control/819_Most_Common_Word/Most_Common_Word.cpp
// 819. Most Common Word
/*
    Given a string paragraph and a string array of the banned words banned, return the most frequent word that is not banned. It is guaranteed there is at least one word that is not banned, and that the answer is unique.

    The words in paragraph are case-insensitive and the answer should be returned in lowercase.

    Example 1:

    Input: paragraph = "Bob hit a ball, the hit BALL flew far after it was hit.", banned = ["hit"]
    Output: "ball"
    Explanation:
    "hit" occurs 3 times, but it is a banned word.
    "ball" occurs twice (and no other word does), so it is the most frequent non-banned word in the paragraph.
    Note that words in the paragraph are not case sensitive,
    that punctuation is ignored (even if adjacent to words, such as "ball,"),
    and that "hit" isn't the answer even though it occurs more because it is banned.
    Example 2:

    Input: paragraph = "a.", banned = []
    Output: "a"

    Constraints:

    1 <= paragraph.length <= 1000
    paragraph consists of English letters, space ' ', or one of the symbols: "!?',;.".
    0 <= banned.length <= 100
    1 <= banned[i].length <= 10
    banned[i] consists of only lowercase English letters.
*/
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iostream>
#include <regex>

std::string mostCommonWord(std::string paragraph, std::vector<std::string> &banned)
{   
    std::regex reg("[!?',;.]");
    paragraph = std::regex_replace(paragraph, reg, " ");
    std::transform(paragraph.begin(), paragraph.end(), paragraph.begin(), ::tolower);
    std::map<std::string, int> common;
    int position = 0;
    int space;
    while (position < paragraph.length()) {
        space  = paragraph.find(' ', position);\
        if (space == position) {
            position = space + 1;
            continue;
        }
        if (space == std::string::npos) {
            std::string&& sub = paragraph.substr(position);
            if(!common.insert(std::make_pair(sub, 1)).second)
                common[sub] += 1;
            break;
        }
        
        std::string&& sub = paragraph.substr(position, space - position);
        if(!common.insert(std::make_pair(sub, 1)).second)
            common[sub] += 1;
        position = space + 1;
    }

    for (std::string& s : banned) {
        common[s] = 0;
    }

    auto res = std::max_element(common.begin(), common.end(), [] (const auto& a, const auto& b) {
        return a.second < b.second;
    });
    
    return res->first;
}

int main() {
    std::string s = "Bob. hIt, baLl";
    std::vector<std::string> banned{"ball"};
    std::cout << mostCommonWord(s, banned);
    return 0;
}